fedbiomed/researcher/requests/_requests.py

```diff
@@ -0,0 +1,495 @@
+    ) -> dict:
+        """Send a training plan and a ApprovalRequest message to node(s).
+
+        If a list of node id(s) is provided, the message will be individually sent
+        to all nodes of the list.
+        If the node id(s) list is None (default), the message is broadcast to all nodes.
+
+        Args:
+            training_plan: the training plan class to send to the nodes for approval.
+            description: Description of training plan approval request
+            nodes: list of nodes (specified by their UUID)
+
+        Returns:
+            a dictionary of pairs (node_id: status), where status indicates to the researcher
+            that the training plan has been correctly downloaded on the node side.
+            Warning: status does not mean that the training plan is approved, only that it has been added
+            to the "approval queue" on the node side.
+        """
+
+        training_plan_instance = training_plan()
+        training_plan_module = 'my_model_' + str(uuid.uuid4())
+        with tempfile.TemporaryDirectory(dir=environ['TMP_DIR']) as tmp_dir:
+            training_plan_file = os.path.join(tmp_dir, training_plan_module + '.py')
+            try:
+                training_plan_instance.save_code(training_plan_file)
+            except Exception as e:
+                logger.error(f"Cannot save the training plan to a local tmp dir : {e}")
+                return {}
+
+            try:
+                _, training_plan_instance = import_class_object_from_file(
+                    training_plan_file, training_plan.__name__)
+                tp_source = training_plan_instance.source()
+            except Exception as e:
+                logger.error(f"Cannot instantiate the training plan: {e}")
+                return {}
+
+        try:
+            minify(tp_source,
+                   remove_annotations=False,
+                   combine_imports=False,
+                   remove_pass=False,
+                   hoist_literals=False,
+                   remove_object_base=True,
+                   rename_locals=False)
+        except Exception as e:
+            # minify does not provide any specific exception
+            logger.error(f"This file is not a python file ({e})")
+            return {}
+
+        # send message to node(s)
+        message = ResearcherMessages.format_outgoing_message({
+            'researcher_id': environ['RESEARCHER_ID'],
+            'description': str(description),
+            'training_plan': tp_source,
+            'command': 'approval'})
+
+        with self.send(message, nodes, policies=[DiscardOnTimeout(5)]) as federated_req:
+            errors = federated_req.errors()
+            replies = federated_req.replies()
+
+            # TODO: Loop over errors and replies
+            for node_id, error in errors.items():
+                logger.info(f"Node ({node_id}) has returned error {error.errnum}, {error.extra_msg}")
+
+        return replies
```

Comment:
Returned value is not what is expected  / described by prototype.
Pushing a fix for this.

Reply:
Moreover, `NodeAgent` is lacking a way to simply request the `node_id` (which is needed for this fix). Thus, one has no easy way to query a `FederatedRequests` nodes IDs (needed for the fix ...)
=> adding this